File: r/trusted.r
source("functions.r")

if (!file.exists("trusted.Rd") | (difftime(Sys.time(), file.info("trusted.Rd")$mtime, units="days") > 1))
   {
   tru <- read.table("../input/gni/trusted.csv", sep=";", h=FALSE, as.is=TRUE, strip.white=TRUE, encoding="UTF-8")
   ifu <- read.table("../input/gni/if.csv", sep=";", h=FALSE, as.is=TRUE, strip.white=TRUE, encoding="UTF-8")
   ifu <- ifu[!duplicated(ifu[,3]),]
   gni <- rbind(tru, ifu)
   trusted <- data.frame(data.source=c(7,8,11), datasetID=c("IF", "ZB", "IPNI"), kingdom=c("Fungi", "Animalia", "Plantae"))
   nz <- read.table("../input/standalone/nz.csv", sep=";", h=TRUE, as.is=TRUE, strip.white=TRUE, encoding="UTF-8")
   union.gni <- data.frame(datasetID=Recode(gni$V1, trusted$data.source, trusted$datasetID),
           taxonID=gni$V2, scientificName=gni$V3, kingdom=Recode(gni$V1, trusted$data.source, trusted$kingdom),
           LocalHomonymFlag=NA)
   union.nz <- data.frame(datasetID=c("NZ"),
           taxonID=nz$uid, scientificName=toupper(nz$name), kingdom=paste("Animalia:", nz$category, sep=""),
           LocalHomonymFlag=nz$homonymFlag)
   all <- rbind(union.gni, union.nz)
   save(all, file="trusted.Rd", compress=TRUE)
   } else
   {
   load("trusted.Rd")
   }

sources <- scan(file="../tmp/trusted_selected.tmp", what="character", quiet=TRUE)
hm <- Hsub(sources[1], sources[-1], all$datasetID, all$scientificName)
column.name <- paste(hm$sublist, "vs", paste(hm$against, collapse="_"), sep="_")
res <- data.frame(all[all$datasetID %in% hm$sublist,], hm$external, InternalHomonymFlag=hm$internal)
names(res)[6] <- tolower(column.name)

# Output
file.name.tgz <- tolower(paste("../output/", column.name, ".tgz", sep=""))
file.name.csv <- "../output/homonym_search_result.csv"
write.table(res, file=file.name.csv, sep="\t", quote=FALSE, col.names=TRUE, row.names=FALSE)
tarfiles <- c("../output/eml.xml", "../output/meta.xml", file.name.csv)
Tar(file.name.tgz, files=tarfiles, compression="gzip", tar="tar")
ifelse(file.exists(file.name.csv), file.remove(file.name.csv))

# ===
# XML output (just in case ;)
# very slow, commented
# library(XML)
# xml <- xmlTree()
# xml$addTag("document", close=FALSE)
# for (i in 1:nrow(res)) {
#    xml$addTag("row", close=FALSE)
#    for (j in names(res)) {
#        xml$addTag(j, res[i, j])
#    }
#    xml$closeTag()
# }
# xml$closeTag()
# file.name <- paste("../output/darwin_core/trusted", paste(hm$sublist, ".xml", sep=""), sep="_")
# write(saveXML(xml), file=tolower(file.name))
# The other one:
# write(file="res.xml", paste("<row>\n", apply(res, 1, function (.x) paste("\t<", names(res), ">", .x, "</", names(res), ">\n", sep="", collapse="")), "</row>\n", sep=""))

File: php/trusted.php
<html>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8">
<body>

<h1>GNI trusted sources processor</h1>

<p>Process any list from GNI trusted source against the given set of other lists, find external and internal homonyms, and supply 2 new columns.

<hr width="30%" align="left">

<form method="post" action="trusted.php">

What to <input type="submit" name="checklist" value="check">&nbsp;:

<input type="radio" name="what" value="IF" /><b>IF</b> &nbsp;&nbsp;OR
<input type="radio" name="what" value="IPNI" /><b>IPNI</b> &nbsp;&nbsp;OR
<input type="radio" name="what" value="ZB" /><b>ZB</b> &nbsp;&nbsp;OR
<input type="radio" name="what" value="NZ" /><b>NZ</b> &nbsp;&nbsp;&mdash;

against&mdash;
<input type="checkbox" name="against[]" value="IF" /><b>IF</b> &nbsp;&nbsp;AND
<input type="checkbox" name="against[]" value="IPNI" /><b>IPNI</b> &nbsp;&nbsp;AND
<input type="checkbox" name="against[]" value="ZB" /><b>ZB</b> &nbsp;&nbsp;AND
<input type="checkbox" name="against[]" value="NZ" /><b>NZ</b> &nbsp;&nbsp;

</form>

<hr width="30%" align="left">

<?php

if (isset($_POST['checklist'])) 
{

$what = $_POST['what'];
$against = $_POST['against'];
$input_file = "../tmp/trusted_selected.tmp";

$fp = fopen($input_file, 'w');
$fw = fwrite($fp, $what . "\n");
foreach ($against as $ag) 
	{
	fwrite($fp, $ag . "\n");
	}
fclose($fp);

$tmp = `cd ../r; Rscript trusted.r`;

}

?>

<p>Results (Darwin Core archive files):<br /><br />

<?php

if ($handle = opendir('../output')) {
    while (false !== ($file = readdir($handle))) {
        if ($file != "." && $file != ".." && (strpos($file, ".tgz") !== false) && (strpos($file, "binomials_") === false)) {
            echo '<a href="../output/'.$file.'">'.$file.'</a><br />';
        }
    }
    closedir($handle);
}

?>

<br><br>
<hr width="30%" align="left">
<hr width="30%" align="left">

<p>The first stage in invisible to user: all trusted sources (generated from GNI and standalone) are unified and merged in one R object (data frame). To be precise, R first checks the file on a disk, and if it is fresh (less then one day old), R reads copy from disk to memory; if not, R creates new object from scratch and writes it to the disk. This R object is used also for uploaded file and texbox processors. Since IF (Index Fungorum) does not contain uninomials, these names are taken from binomials and then unique-ized. Therefore, IF does not contain internal homonyms. IPNI and ZooBank lists contain internal homonyms, but they also contain artifical homonyms which resulted from incomplete names parsing. This situation should improve in a future.

<p>On the second stage, when user clicks on the radio and checkboxes, PHP writes temporary file which contains name of list to compare and names of lists which will be used for comparison. Then R code starts to work with list in order to determine internal and external homonyms in comparison with other lists. The code is simple and fast, and uses R ability to vectorize calculations. In the end, two colums are added to initial list: external homonyms colum (this column has a variable name reflects names of lists which are compared), and internal homonyms ("InternalHomonymFlag"). Both columns are in binary (0/1) format.

<p>Resulted list (with two additional columns) is written to CSV file which is in turn a component of Darwin Core zipped tar archive. Two other files in DC archive (meta.xml and eml.xml) contain meta-information about CSV file. Format of these files will be most probably changed in a future.

</body>
</html>